### DISCUSSION TITLE
all stepgen position-feedback stuck at 0

### DISCUSSION THREAD

**Discussion by unromeo** [Peter-van-Tol/LiteX-CNC#113] [General]
Hi guys, 

You all have my respect for the work you put into this project. 

But please help a noob out here. 

I managed to flash my 5A-75B with litexcnc 1.2.4 using a Raspi 4 and the documentation on readthedocs. 
My json file is with 20 gpio in, 20 gpio out, 2 PWM, 2 encoders and 4 Stepgen. 

I am now trying to test everything out, but I am stuck at jogging the axis, getting joint following errors on all axes. 
Looking at the HAL Pins, I see that none of the stepgen position-feedback pins are changing values, and the position-prediction is around a third of the real value. 
I know I must be missing something basic here, but I need some guidance to figure out what. 

Thanks in advance for helping me out and let me know if you need more info !

![Screenshot_2024-10-21_12-53-22](https://github.com/user-attachments/assets/e2a11898-e02e-4a39-b403-72fb26fa35f6)

[16_25_6_2.json](https://github.com/user-attachments/files/17466369/16_25_6_2.json)
[my-mill_hal.txt](https://github.com/user-attachments/files/17466575/my-mill_hal.txt)
[my-mill_ini.txt](https://github.com/user-attachments/files/17466587/my-mill_ini.txt)

**Comment by Peter-van-Tol** [comment]
I spotted a small problem in you HAL file:
```
addf 5A-75B:V8.0.write servo-thread
addf 5A-75B:V8.0.read servo-thread

setp 5A-75B:V8.0.watchdog.timeout_ns 8000000

loadrt stepgen step_type=0,0,0,0
loadrt pwmgen output_type=1

addf stepgen.capture-position servo-thread
addf motion-command-handler servo-thread
addf motion-controller servo-thread
addf stepgen.update-freq servo-thread
addf pwmgen.update servo-thread

```

The problem is that the `addf` for the read and write functions of the FPGA are right after each other. The other functions therefore don't have time to process the data from the FPGA before the new batch is send. Also, the modules `stepgen` and `pwmgen` are not required, as this is done within the FPGA itself.

The section above should be:
```
# Add functions
addf 5A-75B:V8.0.read servo-thread
addf motion-command-handler servo-thread
addf motion-controller servo-thread
addf 5A-75B:V8.0.write servo-thread

setp 5A-75B:V8.0.watchdog.timeout_ns 8000000
```

Added the modified HAL file for you to try it out.

[my-mill_hal.txt](https://github.com/user-attachments/files/17472594/my-mill_hal.txt)

**Reply by Peter-van-Tol** [reply]
This modification is a quick-win. If it does not give the desired results, please let me know. There might be some other issues causing this, but requiring some more information:
- Which version of LitexCNC did you install. Did you install it from source (GitHub) or with `pip install` (somethings have changed in GitHub, maybe you've found a bug in a not-yet released version)?
- Did you build your firmware on the same system as you are using your CNC machine (to exclude a version mismatch)?

Some other users found using a `pid`  component for the feedback loop is giving better results. However, I don't think that this is the case, because you indicate the values remain `0`, which looks more like a communication error.

**Reply by unromeo** [reply]
and just like that, now it is working as it should ! 

Thanks a lot, learned something new today. 

Cheers !

**Reply by Peter-van-Tol** [reply]
Thanks for your feedback. When you have more questions, feel free to open a new topic.
